functions/snippets/src/test/java/com/example/functions/OcrTests.java

```diff
@@ -0,0 +1,150 @@
+    gcsEvent.name = "wakeupcat.jpg";
+
+    new OcrProcessImage().accept(gcsEvent, null);
+
+    List<LogRecord> logs = LOG_HANDLER.getStoredLogRecords();
+    assertThat(logs.get(1).getMessage()).contains("Extracted text from image: Wake up human!");
+    assertThat(logs.get(2).getMessage()).contains("Detected language en for file wakeupcat.jpg");
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void functionsOcrTranslate_shouldValidateParams() throws IOException {
+    PubSubMessage message = new PubSubMessage();
+    message.data = new String(Base64.getEncoder().encode("{}".getBytes()));
+
+    new OcrTranslateText().accept(message, null);
+  }
+
+  @Test
+  public void functionsOcrTranslate_shouldTranslateText() throws IOException {
+    String text = "Wake up human!";
+    String filename = "wakeupcat.jpg";
+    String lang = "es";
+
+    JsonObject dataJson = new JsonObject();
+    dataJson.addProperty("text", text);
+    dataJson.addProperty("filename", filename);
+    dataJson.addProperty("lang", lang);
+
+    PubSubMessage message = new PubSubMessage();
+    message.data = new String(Base64.getEncoder().encode(gson.toJson(dataJson).getBytes()));
+
+    new OcrTranslateText().accept(message, null);
+
+    List<LogRecord> logs = LOG_HANDLER.getStoredLogRecords();
+    assertThat(logs.get(1).getMessage()).contains("¡Despierta humano!");
+    assertThat(logs.get(2).getMessage()).isEqualTo("Text translated to es");
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void functionsOcrSave_shouldValidateParams() throws IOException {
+    PubSubMessage message = new PubSubMessage();
+    message.data = new String(Base64.getEncoder().encode("{}".getBytes()));
+
+    new OcrSaveResult().accept(message, null);
+  }
+
+  @Test
+  public void functionsOcrSave_shouldPublishTranslatedText() throws IOException {
+    String text = "Wake up human!";
+    String filename = "wakeupcat.jpg";
+    String lang = "es";
+
+    JsonObject dataJson = new JsonObject();
+    dataJson.addProperty("text", text);
+    dataJson.addProperty("filename", filename);
+    dataJson.addProperty("lang", lang);
+
+    PubSubMessage message = new PubSubMessage();
+    message.data = new String(Base64.getEncoder().encode(gson.toJson(dataJson).getBytes()));
+
+    new OcrSaveResult().accept(message, null);
+
+    String resultFilename = "wakeupcat.jpg_to_es.txt";
+
+    // Check log messages
+    List<LogRecord> logs = LOG_HANDLER.getStoredLogRecords();
+    String expectedMessage = String.format(
+        "Saving result to %s in bucket %s", resultFilename, RESULT_BUCKET);
+    assertThat(LOG_HANDLER.getStoredLogRecords().get(1).getMessage()).isEqualTo(expectedMessage);
+
+    // Check for (recently) written file
+    // (Using timestamps instead of exists() saves us a deletion step)
+    BlobInfo resultBlob = storage.get(RESULT_BUCKET, resultFilename);
+    Long fiveMinutesAgo = System.currentTimeMillis() - 60000 * 5;
```

Comment:
Please use `java.time` for this instead.  You can use something like `Instance.now().minus(5, MINUTES)`

Additionally, this check doesn't actually verify it worked when run in parallel (as we do with our unit tests).

Reply:
Replaced this with a UUID'd filename + existence check. 🙂

(The downside is that this also required implementing teardown code.)